### PR TITLE
Add fields to BigQuery tables to link LA DSI users to organisations

### PR DESCRIPTION
**Pull request by david-mears-dfe**
## Jira ticket URL

https://dfedigital.atlassian.net/browse/TEVA-1571

## Changes in this PR

This PR is mainly about updating the job which calls the DSI API and sends the data to BigQuery.

There was an option suggested of scrapping this job, and putting everything in our Users table (and then send that to BigQuery).

I propose keeping the DSI export to BigQuery separate from the User table for two reasons:

1. A user entity in Postgres corresponds to several user entities in BigQuery; in BQ, there is one user record per organisation the user has access to. These differences correspond to different use cases. In the app itself we benefit from consolidating all user data into a single record. In analytics, however, we benefit from keeping them apart so that we can ask questions such as "When did the approvals of users to sign in as LAs happen?" (each BQ user has an 'approved at' field).

2. The principle of not storing data that's only used for analytics/monitoring in the app database (separation of concerns)

I have tested the jobs on the staging dataset.

### CODE REVIEW

**Comment by david-mears-dfe** [comment]
> [](https://sonarcloud.io/component_measures?id=DFE-Digital_teaching-vacancies&pullRequest=2358&metric=new_coverage&view=list) [47.8% Coverage](https://sonarcloud.io/component_measures?id=DFE-Digital_teaching-vacancies&pullRequest=2358&metric=new_coverage&view=list)

This refers to the lines highlighted in red below.

Screenshot 2020-11-17 at 23 12 06

This is an API call. We do in fact test this call, but we stub it out in order to do so.

File: lib/export_dsi_approvers_to_big_query.rb [hunk 1]
```diff
@@ -16,28 +16,30 @@ def run!
...
         email: user["email"],
         family_name: user["familyName"],
         given_name: user["givenName"],
+        la_code: la_code(user),
+        trust_uid: user.dig("organisation", "uid"),
+        role_id: user["roleId"],
+        role_name: user["roleName"],
         school_urn: user.dig("organisation", "urn"),
-        organisation_uid: user.dig("organisation", "uid"),
+        user_id: user["userId"],
```
Comment (david-mears-dfe) [inline lib/export_dsi_approvers_to_big_query.rb:27]: Changes here:
- sort alphabetically
- rename organisation_uid to trust_uid
- add la_code

File: lib/export_dsi_approvers_to_big_query.rb [hunk 2]
```diff
@@ -16,28 +16,30 @@ def run!
...
       schema.string "email", mode: :required
       schema.string "family_name", mode: :required
       schema.string "given_name", mode: :required
+      schema.integer "la_code", mode: :nullable
+      schema.string "role_id", mode: :required
+      schema.string "role_name", mode: :required
       schema.integer "school_urn", mode: :nullable
-      schema.integer "organisation_uid", mode: :nullable
+      schema.integer "trust_uid", mode: :nullable
+      schema.string "user_id", mode: :required
```
Comment (david-mears-dfe) [inline lib/export_dsi_approvers_to_big_query.rb:42]: Changes here:
- sort alphabetically by column
- rename organisation_uid to trust_uid
- add la_code

File: lib/organisation_import/import_organisation_data.rb [hunk 3]
```diff
@@ -4,13 +4,6 @@
 require "open-uri"
 
 class ImportOrganisationData
-# NB: To me it doesn't seem too unreasonable to have the transformation logic here, but in the interest of not
-#     deleting comments I didn't write, see the TODO below.
-# TODO: Refactor the transformation logic into the model.
-# These are the attributes that require additional transformation before being added to the model. The first value of
-# the array is the row key name, the second is the method used for the transformation.  URL is the exception, as it
-# requires an external function call-this is handled in the method.
```
Comment (david-mears-dfe) [inline lib/organisation_import/import_organisation_data.rb:12]: I moved this comment to be next to the code it describes. Look [here](https://github.com/DFE-Digital/teaching-vacancies/blob/9256e4fbc25c85d3bba10618fc2c54551fb14538/lib/update_school_data.rb) to see the original context.

File: app/controllers/hiring_staff/sign_in/dfe/sessions_controller.rb [hunk 4]
```diff
@@ -77,12 +77,17 @@ def trust_uid
...
-    if LocalAuthorityAccessFeature.enabled? && school_urn.blank? && trust_uid.blank?
+    if LocalAuthorityAccessFeature.enabled? && local_authority_sign_in?
       auth_hash.dig("extra", "raw_info", "organisation", "establishmentNumber") || ""
     end
   end
 
+  def local_authority_sign_in?
+    # In organisation['category'], DSI provides an 'id' and a 'name' attribute, mapped as below:
+    # '001' => 'Establishment' (i.e. a school), '002' => 'Local Authority', '010' => 'Multi-Academy Trust'
+    auth_hash.dig("extra", "raw_info", "organisation", "category", "id") == "002"
```
Comment (cpjmcquillan) [inline app/controllers/hiring_staff/sign_in/dfe/sessions_controller.rb:88]: nice 👍

File: spec/fixtures/dfe_sign_in_service_users_response_page_2.json [hunk 5]
```diff
@@ -70,7 +70,7 @@
     "organisation": {
       "id": "13F20E54-79EA-4146-8E39-18197576F024",
       "name": "Organisation 2",
-      "Category": "002",
+      "Category": "010",
```
Comment (alexbowen) [inline spec/fixtures/dfe_sign_in_service_users_response_page_2.json:73]: should this be 001?

File: lib/dfe_sign_in_api.rb [hunk 6]
```diff
@@ -72,4 +72,19 @@ def get_response_pages
...
   end
+
+  def la_code(user)
+    # The organisation in the user from the DSI response contains an establishment number, which
+    # matches the LA code if the organisation is an LA.
+    # Category '002' denotes an LA.
+    # To be consistent with trusts and schools, we only send the LA code to BigQuery if the organisation
+    # is of the right type (i.e. it's an LA).
+
+    # There is a different schema for the approvers and users API responses:
```
Comment (cpjmcquillan) [inline lib/dfe_sign_in_api.rb:83]: Of course there is

**Comment by alexbowen** [comment]
the data in spec/fixtures/dfe_sign_in_service_users_response_page_2.json doesnt look to be quite consistent to me, but im not sure it really matters in context of this

**Review by cpjmcquillan — APPROVED** [review]

File: app/controllers/hiring_staff/sign_in/dfe/sessions_controller.rb (hunk 4 above)
Comment (david-mears-dfe) [reply app/controllers/hiring_staff/sign_in/dfe/sessions_controller.rb]: Moved to another PR

File: spec/fixtures/dfe_sign_in_service_users_response_page_2.json (hunk 5 above)
Comment (david-mears-dfe) [reply spec/fixtures/dfe_sign_in_service_users_response_page_2.json:73]: 010 denotes a trust

**Pull request merged**